pkg/frontend/querymiddleware/astmapper/rangemapper.go

```diff
@@ -0,0 +1,543 @@
+
+// isSplittable returns whether it is possible to optimize the given
+// sample expression.
+// A vector aggregation is splittable, if the aggregation operation is
+// supported and the inner expression is also splittable.
+// A range aggregation is splittable, if the aggregation operation is
+// supported.
+// A binary expression is splittable, if at least one operand is splittable.
+func isSplittable(node parser.Node) bool {
+	switch n := node.(type) {
+	case *parser.AggregateExpr:
+		_, ok := supportedVectorAggregators[n.Op]
+		return ok && isSplittable(n.Expr)
+	case *parser.BinaryExpr:
+		return isSplittable(n.LHS) || isSplittable(n.RHS)
+	case *parser.Call:
+		_, ok := splittableRangeVectorAggregators[RangeVectorName(n.Func.Name)]
+		if ok {
+			return true
+		}
+		var isArgSplittable bool
+		// It is considered splittable if at least a Call argument is splittable
+		for _, arg := range n.Args {
+			isArgSplittable = isSplittable(arg)
+			if isArgSplittable {
+				break
+			}
+		}
+		return isArgSplittable
+	case *parser.ParenExpr:
+		return isSplittable(n.Expr)
+	}
+	return false
+}
+
+func isVectorAggregatorSplittable(expr *parser.AggregateExpr) bool {
+	_, ok := splittableVectorAggregators[expr.Op]
+	return ok
+}
+
+// mapAggregatorExpr maps vector aggregator expression expr
+func (r *rangeMapper) mapAggregatorExpr(expr *parser.AggregateExpr, stats *MapperStats) (mapped parser.Node, finished bool, err error) {
+	// In case the range interval is smaller than the configured split interval,
+	// don't split it
+	rangeInterval := getRangeInterval(expr)
+	if rangeInterval <= r.splitByInterval {
+		return expr, false, nil
+	}
+
+	var mappedNode parser.Node
+
+	// If the embeddedAggregatorExpr is not set, update it
+	// Note: vector aggregators avg, count and topk are supported but not splittable, so cannot be sent downstream
+	if r.embeddedAggregatorExpr == nil && isVectorAggregatorSplittable(expr) {
+		mappedNode, finished, err = NewASTNodeMapper(r.copyWithEmbeddedExpr(expr)).MapNode(expr.Expr, nil)
+	} else {
+		mappedNode, finished, err = r.MapNode(expr.Expr, nil)
+	}
+	if err != nil {
+		return nil, false, err
+	}
+	if !finished {
+		return expr, false, nil
+	}
+
+	mappedExpr, ok := mappedNode.(parser.Expr)
+	if !ok {
+		return nil, false, fmt.Errorf("unable to map expr '%s'", expr)
+	}
+
+	// Create the parent expression while preserving the grouping from the original one
+	return &parser.AggregateExpr{
+		Op:       expr.Op,
+		Expr:     mappedExpr,
+		Param:    expr.Param,
+		Grouping: expr.Grouping,
+		Without:  expr.Without,
+	}, true, nil
+}
+
+// mapBinaryExpr maps binary expression expr
+func (r *rangeMapper) mapBinaryExpr(expr *parser.BinaryExpr, stats *MapperStats) (mapped parser.Node, finished bool, err error) {
+	// Noop if both LHS and RHS are literal numbers
+	_, literalLHS := expr.LHS.(*parser.NumberLiteral)
+	_, literalRHS := expr.RHS.(*parser.NumberLiteral)
+	if literalLHS && literalRHS {
+		return expr, false, nil
+	}
```

Comment:
Can we reach this point? I think we shouldn't as `isSplittable` should have returned `false` already.

I'm okay with having a check here, but I'd return an error if we reach this point.

Reply:
Ah yes, you are right, isSplittable would not allow this to happen. Will return an error 👍